ipa-core/src/protocol/basics/mul/dzkp_malicious.rs

```diff
@@ -113,7 +113,7 @@ mod test {
                 assert!(matches!(mctx.is_verified(), Err(Error::ContextUnsafe(_))));
 
                 // validate all elements in the batch
-                validator.validate().await.unwrap();
+                validator.validate(0usize).await.unwrap();
```

Comment:
This is kind of weird. If you aren't calling the `validated_seq_join` version, passing a `0usize` here is kind of strange.

Elsewhere, the way we approach this is to first narrow the context. Might there be some similar solution here?

Reply:
I agree. I have tried to put the counter as a field in the validator, but then validate needs to take &mut self rather than &self which causes problems with FnMut closures.

Reply:
There is a `Mutex` within the validator if you go deep enough, so it seems like it ought to be possible to leverage that somehow to get interior mutability without a major change to the contract / behavior. But it might take a little rearranging of the pieces.

Reply:
@andyleiserson so you suggest to store the counter on the batch? 

```
pub struct MaliciousDZKPValidator<'a> {
    batch_ref: Arc<Mutex<Batch>>,
    protocol_ctx: MaliciousDZKPUpgraded<'a>,
    validate_ctx: Base<'a>,
}
```

I think that would a bit strange since the counter is used to narrow down `validate_ctx` which is also not stored in the batch.